arcgis-ios-sdk-samples/Utility Network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift

```diff
@@ -0,0 +1,271 @@
+// Copyright 2020 Esri
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+import ArcGIS
+
+class PerformValveIsolationTraceViewController: UIViewController {
+    // MARK: Storyboard views
+    
+    /// The button to start the trace.
+    @IBOutlet weak var traceButton: UIBarButtonItem!
+    /// The button to choose a utility category for filter barriers.
+    @IBOutlet weak var categoryButton: UIBarButtonItem!
+    /// The switch to control whether to include isolated features in the trace results when used in conjunction with an isolation trace.
+    @IBOutlet weak var isolationSwitch: UISwitch!
+    /// The label to display trace status.
+    @IBOutlet weak var statusLabel: UILabel!
+    /// The map view managed by the view controller.
+    @IBOutlet weak var mapView: AGSMapView! {
+        didSet {
+            layers = makeLayers(from: featureServiceURL)
+            mapView.map = makeMap(with: layers)
+        }
+    }
+    
+    // MARK: Instance properties
+    
+    /// The URL to the feature service for running the isolation trace.
+    let featureServiceURL = URL(string: "https://sampleserver7.arcgisonline.com/arcgis/rest/services/UtilityNetwork/NapervilleGas/FeatureServer")!
+    
+    // Constants for creating the default trace configuration.
+    let domainNetworkName = "Pipeline"
+    let tierName = "Pipe Distribution System"
+    
+    // Constants for creating the default starting location.
+    let networkSourceName = "Gas Device"
+    let assetGroupName = "Meter"
+    let assetTypeName = "Customer"
+    let globalId = UUID(uuidString: "98A06E95-70BE-43E7-91B7-E34C9D3CB9FF")!
+    
+    /// An array to keep track of all available utility categories under current network definition.
+    var filterBarrierCategories: [AGSUtilityCategory]!
+    
+    /// The feature layers created from the feature service URL.
+    var layers: [AGSFeatureLayer]!
```

Comment:
There's no need for this to be an implicitly unwrapped optional. We try to avoid optional collections whenever possible since, in almost all cases, `nil` and `isEmpty == true` mean the same thing. You could either initialize this with the layers by calling `makeLayers(from:)`:

```swift
lazy var layers: [AGSFeatureLayer] = {
    return self.makeLayers(from: featureServiceURL)
}()
```

Or just combine the two:

```swift
lazy var layers: [AGSFeatureLayer] = {
    let urls = [
        featureServiceURL.appendingPathComponent("3"),
        featureServiceURL.appendingPathComponent("0")
    ]
    return urls.map { (url) -> AGSFeatureLayer in
        let featureTable = AGSServiceFeatureTable(url: url)
        return AGSFeatureLayer(featureTable: featureTable)
    }
}()
```

And, either way, eliminate the assignment on line `32`.

Reply:
This is better 👍